In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Data

In [24]:
data = pd.read_csv('./data.csv')
data

Title         Years Certification  Runtime  \
0                     Queen Cleopatra  (2023, None)         TV-14     45.0   
1                               Velma  (2023, 2024)         TV-MA     25.0   
2     Keeping Up with the Kardashians  (2007, 2021)         TV-14     44.0   
3                            Batwoman  (2019, 2022)         TV-14     45.0   
4                         The Acolyte  (2024, None)         TV-14     35.0   
...                               ...           ...           ...      ...   
1133                         The Wire  (2002, 2008)         TV-MA     60.0   
1134                     Planet Earth  (2006, None)         TV-PG     50.0   
1135                 Band of Brothers  (2001, None)         TV-MA     60.0   
1136                  Planet Earth II  (2016, None)          TV-G     50.0   
1137                     Breaking Bad  (2008, 2013)         TV-MA     45.0   

      Rating  Number of Votes  Emmys                       Creators  \
0        1.2            86000      0                            NaN   
1        1.6            80000      0                 Charlie Grandy   
2        2.9            32000      0  Ryan Seacrest, Eliot Goldberg   
3        3.6            47000      0                 Caroline Dries   
4        4.1           125000      0                Leslye Headland   
...      ...              ...    ...                            ...   
1133     9.3           390000      0                    David Simon   
1134     9.4           223000      4                            NaN   
1135     9.4           544000      6                            NaN   
1136     9.5           162000      2                            NaN   
1137     9.5          2200000     16                 Vince Gilligan   

                                                 Actors  \
0     Jada Pinkett Smith, Adele James, Craig Russell...   
1     Mindy Kaling, Glenn Howerton, Sam Richardson, ...   
2     Khloé Kardashian, Kim Kardashian, Kourtney Kar...   
3     Camrus Johnson, Rachel Skarsten, Meagan Tandy,...   
4     Lee Jung-jae, Amandla Stenberg, Manny Jacinto,...   
...                                                 ...   
1133  Dominic West, Lance Reddick, Sonja Sohn, Wende...   
1134  Sigourney Weaver, David Attenborough, Nikolay ...   
1135  Scott Grimes, Damian Lewis, Ron Livingston, Sh...   
1136  David Attenborough, Michael J. Sanderson, Gord...   
1137  Bryan Cranston, Aaron Paul, Anna Gunn, Betsy B...   

                                                 Genres  \
0                           Documentary, Drama, History   
1     Animation, Adventure, Comedy, Crime, Horror, M...   
2                                    Family, Reality-TV   
3               Action, Adventure, Crime, Drama, Sci-Fi   
4     Action, Adventure, Drama, Fantasy, Mystery, Sc...   
...                                                 ...   
1133                             Crime, Drama, Thriller   
1134                                Documentary, Family   
1135                                Drama, History, War   
1136                                        Documentary   
1137                             Crime, Drama, Thriller   

                                    Coutries of origins  \
0                                        United Kingdom   
1                            United States, South Korea   
2                                         United States   
3                                         United States   
4                                         United States   
...                                                 ...   
1133                                      United States   
1134       United Kingdom, Canada, United States, Japan   
1135                      United Kingdom, United States   
1136  United Kingdom, Germany, France, China, United...   
1137                                      United States   

                                       Languages  \
0                                        English   
1          

# Train Model

In [25]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error



X = data[['Runtime', 'Number of Votes', 'Emmys']]
y = data['Rating']  

# Chia tập train-test (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Khởi tạo và huấn luyện mô hình RandomForest
model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

# Dự đoán trên tập kiểm tra
y_pred = model.predict(X_test)

# Đánh giá mô hình
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"Mean Absolute Error (MAE): {mae:.4f}")
print(f"R² của mô hình RandomForest: {r2:.4f}")

# Lấy tầm quan trọng của các đặc trưng (feature importances)
feature_importances = pd.DataFrame({
    'Feature': X.columns,
    'Importance': model.feature_importances_
}).sort_values(by='Importance', ascending=False)


print("Tầm quan trọng của các đặc trưng:")
print(feature_importances)


Mean Squared Error (MSE): 0.5765
Mean Absolute Error (MAE): 0.5616
R² của mô hình RandomForest: 0.0484
Tầm quan trọng của các đặc trưng:
           Feature  Importance
1  Number of Votes    0.540540
0          Runtime    0.382493
2            Emmys    0.076967


We notice that the R² value of the RandomForest model is only 0.0484, which is very low compared to expectations. R² (the coefficient of determination) is an important metric for evaluating how well the model fits the actual data. The closer the R² value is to 1.0, the better the model explains the variability in the data. Conversely, a value close to 0, like in this case, indicates that the model explains only a very small portion of the data's variability.

With an R² value of 0.0484, this suggests that the RandomForest model performs very poorly in predicting or explaining the changes in the data. Specifically, the model explains only about 4.84% of the data's variability, meaning that it has not captured the important relationships between the factors in the dataset.

Here, we observe that two columns have a significant impact on the **Rating**, which are **Number of Votes** and **Runtime**. Specifically, **Number of Votes** has a correlation coefficient of **0.540540**, indicating that the more votes a movie receives, the higher its rating tends to be. This reflects the idea that movies with a larger viewership and more voter participation often receive higher ratings.

Additionally, **Runtime** also has a notable influence, with a correlation coefficient of **0.382493**. This suggests that the movie's duration plays an important role in shaping the viewer's rating. Movies with an appropriate runtime—neither too long nor too short—tend to receive higher ratings, as viewers may feel more comfortable while watching them. These factors together contribute to the overall rating given by the audience for the films.

# Optimization: 
<br>You can add some values such as:
- **n_estimators**: The number of trees in the forest. Increasing this value can improve the model's performance but also increases computation time. However, after a certain number of trees, the performance may not change much.
- **max_depth**: The maximum depth of each tree. Limiting the depth can help reduce overfitting.
- **min_samples_split**: The minimum number of samples required to split a node. Increasing this value can help reduce model complexity.
- **min_samples_leaf**: The minimum number of samples required at each leaf. This helps reduce overfitting by avoiding overly detailed splits.
- **max_features**: The maximum number of features to consider at each split. Limiting the number of features can help reduce model complexity.
- **bootstrap**: Specifies whether to use bootstrap sampling (sampling with replacement). Try both True and False to see the effect on the results.